### ISSUE TITLE
RFC - Upcoming a11y design

### ISSUE THREAD

**Issue by cibernox** [cibernox/ember-power-select#293]
After some comments in other issues, I've decided to gather together all the a11y design and best practices. 

I've also found [this example](https://www.w3.org/TR/wai-aria/roles#combobox) to be particularly enlightening. 

NOTE: The image deliberately omit the roles of the options of the list because seems to be a consensus that are `role=option aria-selected=true/false aria-current=true/false aria-disabled=true/false`. This focuses on what wires the trigger, the searbox and the list of options together only.

Starting with the most basic component, Ember Basic Dropdown, the a11y is relatively simple.

<img width="706" alt="basic-dropdown-a11y" src="https://cloud.githubusercontent.com/assets/265339/12863279/b067a564-cc6b-11e5-96a1-7e1a80772533.png">

On top of this we build the power-select and power-select-multiple

<img width="689" alt="single-select-a11y" src="https://cloud.githubusercontent.com/assets/265339/12863320/c71ad934-cc6b-11e5-976d-2fa6902fc403.png">

<img width="627" alt="multiple-select-a11y" src="https://cloud.githubusercontent.com/assets/265339/12863323/d0988aec-cc6b-11e5-85d3-3e9d370378b1.png">

I'm also omitting `aria-description`, `aria-label`.

cc @jmacqueen @martndemus

In case this is the path taken:

- [x] The options list has `role=listbox`.
- [x] Groups have `role=group`
- [x] Each option has `role=option`.
- [x] The selected option has `aria-selected=true`.
- [x] The highlighted option has `aria-current=true`.
- [x] The disabled options has `aria-disabled=true`.
- [x] Searchbox of the single component has `role=combobox`.
- [ ] Searchbox of the single select has `aria-expanded=true`, `aria-owns=id-of-the-listbox` and `aria-autocomplete=list`.
- [ ] Searchbox of the multiple select has `aria-expanded=true`, `aria-owns=id-of-the-listbox` and `aria-autocomplete=list`. **aria-expanded** has to be false/true depending on if the options are visible?
- [ ] Trigger has `role=button`
- [x] Trigger `aria-owns=<id-of-dropdown>`
- [ ] Trigger has `aria-pressed=true/false`. **Should we remove `aria-expanded` then?**
- [ ] Trigger get `aria-disabled=true` when passed `disabled=true` to the component.
- [ ] Trigger has `aria-haspopup=true`.
- [ ] Accepts `aria-labelledby` (applies to the trigger).
- [ ] Accepts `aria-describedby` (applies to the trigger)
- [ ] Accepts `aria-requited` (applies to the trigger)
- [ ] Accepts `aria-invalid` (applies to the trigger)

**Comment by cibernox** [comment]
The main point of this design for me is that is consistent. 
There is the `button` + `dropdown` responsibility, owned by ember-basic-dropdown, and then there is the `combobox` + `listbox` responsibility that lives in Ember Power Select. They are 2 reparate concerns.

**Comment by jmacqueen** [comment]
I think we're working really hard to pound a square peg into a round hole.

`aria-autocomplete` and `combobox` semantics are very specific. They expect that the ultimate result will reside in whatever element is flagged as a `combobox` and `aria-autocomplete=list` also *requires* the use of `aria-activedescendant`.

That's not where the search results live here, which is why I was suggesting in the other thread that we dispense with `role=combobox` all together and leverage `aria-controls` instead. Ultimately, I think it will be a simpler way to get something validly accessible. By my reading, what you have here is going to be so hard to represent as a `combobox` that you might as well skip it and just lay out the controlling relationships between the pieces.

**Comment by jmacqueen** [comment]
I think the dropdown semantics are pretty well understood (my awkward attempt to diagram follows)
![dropdown aria 001](https://cloud.githubusercontent.com/assets/5197963/12889434/dc6e3d52-ce4c-11e5-9fdb-0c55c83df0e3.png)

 - [ ]

**Comment by cibernox** [comment]
@jmacqueen Yes, on the dropdown I think the looks ok.
Should se use `aria-owns` or `aria-controls` (or both)?

**Comment by jmacqueen** [comment]
I prematurely sent the earlier post. It's updated now. But the answer is `aria-controls` is all you need.

More amazing diagrams to follow.

**Comment by jmacqueen** [comment]
Single Select key points

<img width="803" alt="screen shot 2016-02-08 at 10 40 17 am" src="https://cloud.githubusercontent.com/assets/5197963/12890138/5fd38f64-ce50-11e5-8fd0-52430d758201.png">

#### Search
  - [ ] `<input type=search>`
  - [ ] `aria-controls=<id of listbox>`

#### Listbox
  - [ ] `role=listbox`
  - [ ] `aria-controls=<id of trigger>` since the trigger will hold the content changed from the `listbox`

#### Options
All the usual `option` stuff

**Comment by jmacqueen** [comment]
I'm looking at the multiple select now. I have a little time before I get pulled into more actual work :smile:

**Comment by MichielBijl** [comment]
Please see the [Authoring Practices Guide's section on autocomplete](http://w3c.github.io/aria/practices/aria-practices.html#autocomplete) for more information about keyboard handling.

ps. there is a discussions going on at the moment regarding combobox:

[ACTION-1490 proposal says good bye to the "inline" notion for combobox](https://lists.w3.org/Archives/Public/public-aria/2016Feb/0105.html)

**Comment by jmacqueen** [comment]
## Multi-select key points
First pass

<img width="837" alt="screen shot 2016-02-08 at 1 38 34 pm" src="https://cloud.githubusercontent.com/assets/5197963/12895223/4a6f211a-ce69-11e5-9df2-50b3483b603c.png">

#### In the trigger
 - [ ] **Search Box**
  - [ ] `<input type="search">`
  - [ ] `aria-controls=<id of listbox>`
  - [ ] include **close button**
 - [ ] **List of selections** `role=list` (this is different from `listbox` in that the individual items are not expected to be directly clickable). `<ul>` has an implicit `role=list`
  - [ ] **Selection** with `role=listitem`. `<li>` has an implicit `role=listitem`
    - [ ] **Close Button** with `role=button` or preferably an actual `<button>`
      - [ ] `aria-label=something useful `
      - [ ] `tabindex=0` to allow each close button to receive focus

#### In the dropdown
 - [ ] **Listbox**
  - [ ] Basically the same as single select, but ideally, `aria-controls=<id of selection list>`

**Comment by cibernox** [comment]
@jmacqueen I've implemented all tasks but two:

* `tabindex=0` to remove buttons. I will need some more work
* Search Box: include close button < -- I don't actually understand this

**Comment by miguelcobain** [comment]
@cibernox I see the value in using aria attributes for styling. However, css attribute selectors are known to perform badly in IE browsers. I'm not saying this by my own experience, unfortunately.

I've seen that angular material recently changed their layout styles to use class selectors instead of attribute selectors. See https://material.angularjs.org/latest/layout/introduction in the section "under-the-hood".
Angular Material is targeted for all browsers with versions n-1; where n is the current browser version, so I assume this was a real problem they've faced?

**Comment by jmacqueen** [comment]
@cibernox "Search Box: include close button" means that the search box in use in multi-select also has its own close button and it would have the same structure as the close button in each of the selections. Sorry that was unclear.

@miguelcobain From my reading ([insert first thing that Google put up here as an example](https://benfrain.com/css-performance-revisited-selectors-bloat-expensive-styles/)), I haven't seen any benchmarking of attribute selectors that shows an appreciable difference in performance unless you have well over 20,000 elements on the page. This coupled with the WCAG recommendation to style against the WAI-ARIA attributes wherever possible [ARIA Authoring Practices](https://www.w3.org/TR/wai-aria-practices/) (see number 7). It has problems in IE6, but that's not even supported by Microsoft any more and I'm well past caring about it :smile:

**Comment by miguelcobain** [comment]
@jmacqueen more context:
- https://github.com/angular/material/pull/4282
- https://github.com/angular/material/issues/1771
- I believe this was the point where they acknowledged the perf problem: https://github.com/angular/material/issues/1771#issuecomment-133268555

I'm not sure this results in a significant perf penalty, but it might be worth to try to quantify that penalty.

**Comment by jmacqueen** [comment]
Ah. After looking at what they were replacing. That project was using a huge number of unqualified attribute selectors, which is the equivalent of doing a global select and then an attribute select on every rule. Very, very, very slow combination. And they had a TON of them in the [layout module](https://github.com/angular/material/blob/release/0.10.0/src/core/style/layout.scss).

As long as there is some qualifier (e.g. `ember-basic-dropdown-trigger[aria-expanded="true"]`) then there shouldn't be any problem performance-wise.

**Comment by miguelcobain** [comment]
@jmacqueen got it! Thanks for the explanation.

**Comment by cibernox** [comment]
@jmacqueen Thanks for investigating that. I never considered css selector performance as a critical thing before.

Some downside of of using selectors like `.ember-power-select-option[aria-dirabled="true"]` is that this is a 2nd level selector, and has higher precedence than `.ember-power-select-option--selected`.

I could have mantained the classes but I didn't because 
1) Avoid redundancy 
2) Performance (The least bindings ember has to maintain, the best)
3) Enforce a11y on anyone that wants to build their custom components but want to reuse styles.

Let's see how it goes.

**Comment by cibernox** [comment]
@jmacqueen 
> Search Box: include close button

Mmm.. there is no close button. I think that the button you see in chrome is because it's added by default by blink and webkit to `input[type=search]` things. Is that what you mean?

**Comment by jmacqueen** [comment]
Hah! Yes! That's exactly what I mean. Oops.

In that case, it needs no special treatment. The browser will handle populating the AT (as it does for any other built-in interactive element).

**Comment by jmacqueen** [comment]
re: CSS...how are you limiting your `aria-disabled` styling to be specific to `power-select` so it doesn't hit everything on the page with `aria-disabled`?

**Comment by cibernox** [comment]
I think I'm only targeting aria-disabled on the options and the trigger.

**Comment by jmacqueen** [comment]
Is there a way to pass through `aria-labelledby`, `aria-describedby`, `aria-required`, `aria-invalid`, and/or `aria-label` from the `{{power-select}}` invocation into the `{{basic-dropdown}}`?

**Comment by cibernox** [comment]
@jmacqueen PR #254 should be resuscitated for that. @martndemus are you still interested in finish that?

**Comment by martndemus** [comment]
@cibernox Not really sure when I can commit to this, but if I have some time I sure would like to help :smile:

**Comment by jmacqueen** [comment]
Looking at it now. I think I was might have been unclear with my dropdown diagram back in February. 

My intention was for `role=button`, `aria-haspopup=true`, `aria-expanded=true/false`, `aria-pressed=true/false`, and `aria-controls=<id of dropdown>` to be on the `ember-basic-dropdown-trigger` `<div>`, not the contained `ember-power-select` trigger. All of those semantics are particular to the dropdown state and best handled at that level...and duplicating that work should be unnecessary.
